File: pulsar-ql-server/src/main/java/fun/platonic/pulsar/ql/h2/udfs/CommonFunctions.java
package fun.platonic.pulsar.ql.h2.udfs;

import fun.platonic.pulsar.ql.DbSession;
import fun.platonic.pulsar.ql.QueryEngine;
import fun.platonic.pulsar.ql.QuerySession;
import fun.platonic.pulsar.ql.annotation.UDFGroup;
import fun.platonic.pulsar.ql.annotation.UDFunction;
import fun.platonic.pulsar.common.NetUtil;
import fun.platonic.pulsar.common.RegexExtractor;
import fun.platonic.pulsar.common.SParser;
import fun.platonic.pulsar.common.URLUtil;
import fun.platonic.pulsar.common.config.ImmutableConfig;
import fun.platonic.pulsar.common.proxy.ProxyEntry;
import fun.platonic.pulsar.common.proxy.ProxyPool;
import fun.platonic.pulsar.persist.metadata.BrowserType;
import fun.platonic.pulsar.persist.metadata.FetchMode;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.Pair;
import org.h2.engine.Session;
import org.h2.ext.pulsar.annotation.H2Context;
import org.h2.tools.SimpleResultSet;
import org.h2.value.Value;
import org.h2.value.ValueArray;
import org.h2.value.ValueString;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import java.io.File;
import java.time.Duration;
import java.util.Map;
import java.util.Objects;

import static fun.platonic.pulsar.common.config.CapabilityTypes.*;
import static java.lang.Long.MAX_VALUE;

@SuppressWarnings("unused")
@UDFGroup
public class CommonFunctions {

    public static final Logger LOG = LoggerFactory.getLogger(CommonFunctions.class);

    private static QueryEngine engine;
    private static ImmutableConfig immutableConfig;
    static {
        engine = QueryEngine.getInstance();
        immutableConfig = engine.getConf();
    }

    /**
     * Set config to the given value
     * @param h2session H2 session
     * @param mode The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     * */
    @UDFunction
    public static String setFetchMode(@H2Context Session h2session, String mode, int ttl) {
        try {
            mode = FetchMode.valueOf(mode.toUpperCase()).toString();
        } catch (Throwable e) {
            LOG.warn("Unknown FetchMode " + mode);
            mode = null;
        }

        LOG.debug("Set fetch mode to " + mode);
        return getAndSetConf(h2session, FETCH_MODE, mode, ttl);
    }

    @UDFunction
    public static String unsetFetchMode(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(FETCH_MODE);
    }

    /**
     * Set config to the given value
     * @param h2session H2 session
     * @param browser The browser to use
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     * */
    @UDFunction
    public static String setBrowser(@H2Context Session h2session, String browser, int ttl) {
        BrowserType browserType = BrowserType.fromString(browser);
        if (browserType == BrowserType.NATIVE) {
            unsetFetchMode(h2session);
        } else {
            setFetchMode(h2session, FetchMode.SELENIUM.name(), ttl);
        }

        LOG.debug("Set browser to " + browser);
        return getAndSetConf(h2session, SELENIUM_BROWSER, browserType.name(), ttl);
    }

    @UDFunction
    public static String unsetBrowser(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        unsetFetchMode(h2session);
        return session.getConfig().getAndUnset(SELENIUM_BROWSER);
    }

    /**
     * Set config to the given value
     * @param h2session H2 session
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     * */
    @UDFunction
    public static String setParallelFetch(@H2Context Session h2session, boolean parallel, int ttl) {
        return getAndSetConf(h2session, FETCH_PREFER_PARALLEL, Boolean.toString(parallel), ttl);
    }

    @UDFunction
    public static String unsetParallelFetch(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(FETCH_PREFER_PARALLEL);
    }

    /**
     * Also retrieve faded links whenever get out links of a page
     * @param ttl The property value time to live in session
     * */
    @UDFunction
    public static String enableFadedLinks(@H2Context Session h2session, int ttl) {
        return getAndSetConf(h2session, PARSE_RETRIEVE_FADED_LINKS, Boolean.toString(true), ttl);
    }

    @UDFunction
    public static String disableFadedLinks(@H2Context Session h2session, int ttl) {
        return getAndSetConf(h2session, PARSE_RETRIEVE_FADED_LINKS, Boolean.toString(false), ttl);
    }

    @UDFunction
    public static String setEagerFetchLimit(@H2Context Session h2session, boolean parallel, int ttl) {
        return getAndSetConf(h2session, FETCH_EAGER_FETCH_LIMIT, Boolean.toString(parallel), ttl);
    }

    @UDFunction
    public static String unsetEagerFetchLimit(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(FETCH_EAGER_FETCH_LIMIT);
    }

    /**
     * Set page expires, this config affects just in h2session scope
     * SQL examples:
     * <code>CALL setPageExpires('1d')</code>
     * <code>CALL setPageExpires('PT24H')</code>
     * <code>CALL setPageExpires('1s')</code>
     * <code>CALL setPageExpires('1')</code>
     *
     * @param h2session  The H2 session, auto injected by h2 runtime
     * @param duration The duration string.
     *                 Supported formats: 1. number in seconds, 2. ISO-8601 standard and 3. hadoop time duration format
     *                 ISO-8601 standard : PnDTnHnMn.nS
     *                 Hadoop time duration format : Valid units are : ns, us, ms, s, m, h, d.
     * @param ttl The property value time to live in session
     * @return The old page expires in seconds, or null if failed
     */
    @UDFunction
    public static String setPageExpires(@H2Context Session h2session, String duration, int ttl) {
        Duration d = getDuration(duration);
        String value = d != null ? d.toString() : null;
        return getAndSetConf(h2session, STORAGE_DATUM_EXPIRES, value, ttl);
    }

    @UDFunction
    public static String unsetPageExpires(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(STORAGE_DATUM_EXPIRES);
    }

    /**
     * Set config to the given value
     * @param h2session H2 session
     * @param duration The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     * */
    @UDFunction
    public static String setPageLoadTimeout(@H2Context Session h2session, String duration, int ttl) {
        Duration d = getDuration(duration);
        String value = d != null ? d.toString() : null;
        return getAndSetConf(h2session, FETCH_PAGE_LOAD_TIMEOUT, value, ttl);
    }

    @UDFunction
    public static String unsetPageLoadTimeout(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(FETCH_PAGE_LOAD_TIMEOUT);
    }

    /**
     * Set config to the given value
     * @param h2session H2 session
     * @param duration The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     * */
    @UDFunction
    public static String setScriptTimeout(@H2Context Session h2session, String duration, int ttl) {
        Duration d = getDuration(duration);
        String value = d != null ? d.toString() : null;
        return getAndSetConf(h2session, FETCH_SCRIPT_TIMEOUT, value, ttl);
    }

    @UDFunction
    public static String unsetScriptTimeout(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(FETCH_SCRIPT_TIMEOUT);
    }

    /**
     * Set config to the given value
     * @param h2session H2 session
     * @param count The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     * */
    @UDFunction
    public static String setScrollDownCount(@H2Context Session h2session, int count, int ttl) {
        return getAndSetConf(h2session, FETCH_SCROLL_DOWN_COUNT, Integer.toString(count), ttl);
    }

    @UDFunction
    public static String unsetScrollDownCount(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(FETCH_SCROLL_DOWN_COUNT);
    }

    /**
     * Set config to the given value
     * @param h2session H2 session
     * @param duration The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     * */
    @UDFunction
    public static String setScrollDownWait(@H2Context Session h2session, String duration, int ttl) {
        Duration d = getDuration(duration);
        String value = d != null ? d.toString() : null;
        return getAndSetConf(h2session, FETCH_SCROLL_DOWN_WAIT, value, ttl);
    }

    @UDFunction
    public static String unsetScrollDownWait(@H2Context Session h2session) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(FETCH_SCROLL_DOWN_WAIT);
    }

    /**
     * Set h2session scope configuration.
     * <p>
     * SQL examples:
     * <code>CALL setConf('fetch.page.load.timeout', '1d')</code>
     * <code>CALL setConf('fetch.fetch.mode', 'NATIVE')</code>
     * <p>
     *
     * @param h2session The H2 session, auto injected by h2 runtime
     * @return The old value of the key
     */
    @UDFunction
    public static String setConfig(@H2Context Session h2session, String name, String value, int ttl) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndSet(name, value, ttl);
    }

    @UDFunction
    public static String setConfig(@H2Context Session h2session, String name, String value) {
        return setConfig(h2session, name, value, Integer.MAX_VALUE / 2);
    }

    @UDFunction
    public static String setConf(@H2Context Session h2session, String name, String value, int ttl) {
        return setConfig(h2session, name, value, ttl);
    }

    @UDFunction
    public static String setConf(@H2Context Session h2session, String name, String value) {
        return setConfig(h2session, name, value, Integer.MAX_VALUE / 2);
    }

    @UDFunction
    public static String unsetConf(@H2Context Session h2session, String name) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(name);
    }

    @UDFunction
    public static String unsetConfig(@H2Context Session h2session, String name) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().getAndUnset(name);
    }

    @UDFunction
    public static String getConf(@H2Context Session h2session, String name) {
        QuerySession session = engine.getSession(new DbSession(h2session));
        return session.getConfig().get(name);
    }

    @UDFunction
    public static String getConfig(@H2Context Session h2session, String name) {
        return getConf(h2session, name);
    }

    @UDFunction
    public static SimpleResultSet config() {
        SimpleResultSet rs = new SimpleResultSet();

        rs.addColumn("NAME");
        rs.addColumn("VALUE");

        for (Map.Entry<String, String> entry : immutableConfig.unbox()) {
            rs.addRow(entry.getKey(), entry.getValue());
        }

        return rs;
    }

    @UDFunction
    public static SimpleResultSet sessionConfig(@H2Context Session h2session) {
        SimpleResultSet rs = new SimpleResultSet();
        rs.addColumn("NAME");
        rs.addColumn("VALUE");

        ImmutableConfig conf = engine.getSession(new DbSession(h2session)).getConfig();
        for (Map.Entry<String, String> entry : conf.unbox()) {
            rs.addRow(entry.getKey(), entry.getValue());
        }

        return rs;
    }

    @UDFunction
    public static SimpleResultSet sysInfo(@H2Context Session h2session) {
        SimpleResultSet rs = new SimpleResultSet();
        rs.addColumn("NAME");
        rs.addColumn("VALUE");

        /* Total number of processors or cores available to the JVM */
        rs.addRow("Available processors (cores)", Runtime.getRuntime().availableProcessors());

        /* Total amount of free memory available to the JVM */
        rs.addRow("Free memory (Mbytes)", Runtime.getRuntime().freeMemory() / 1024 / 1024);

        /* This will return Long.MAX_VALUE if there is no preset limit */
        long maxMemory = Runtime.getRuntime().maxMemory();
        /* Maximum amount of memory the JVM will attempt to use */
        rs.addRow("Maximum memory (Mbytes)", (maxMemory == MAX_VALUE) ? "no limit" : Long.toString(maxMemory / 1024 / 1024));

        /* Total memory currently available to the JVM */
        rs.addRow("Total memory available to JVM (Mbytes)", Runtime.getRuntime().totalMemory() / 1024 / 1024);

        /* Get a list of all filesystem roots on this system */
        File[] roots = File.listRoots();
        for (File root : roots) {
            /* For each filesystem root, print some info */
            rs.addRow("File system root", root.getAbsolutePath());
            rs.addRow("Total space (Mbytes)", root.getTotalSpace() / 1024 / 1024);
            rs.addRow("Free space (Mbytes)", root.getFreeSpace() / 1024 / 1024);
            rs.addRow("Usable space (Mbytes)", root.getUsableSpace() / 1024 / 1024);
        }

        return rs;
    }

    @UDFunction
    public static boolean addProxy(String ipPort) {
        ProxyEntry proxyEntry = ProxyEntry.parse(ipPort);
        if (proxyEntry != null && proxyEntry.testNetwork()) {
            ProxyPool proxyPool = ProxyPool.getInstance(immutableConfig);
            return proxyPool.offer(proxyEntry);
        }

        return false;
    }

    @UDFunction
    public static boolean addProxy(String ip, int port) {
        if (NetUtil.testNetwork(ip, port)) {
            ProxyPool proxyPool = ProxyPool.getInstance(immutableConfig);
            return proxyPool.offer(new ProxyEntry(ip, port));
        }

        return false;
    }

    @UDFunction
    public static int addProxies(ValueArray ipPorts) {
        int count = 0;

        for (Value value : ipPorts.getList()) {
            if (addProxy(value.getString())) {
                ++count;
            }
        }

        return count;
    }

    @UDFunction
    public static int addProxiesUnchecked(ValueArray ipPorts) {
        int count = 0;

        ProxyPool proxyPool = ProxyPool.getInstance(immutableConfig);
        for (Value value : ipPorts.getList()) {
            ProxyEntry proxyEntry = ProxyEntry.parse(value.getString());
            if (proxyEntry != null) {
                proxyPool.offer(proxyEntry);
                ++count;
            }
        }

        return count;
    }

    @UDFunction
    public static int recoverProxyPool(int n) {
        ProxyPool proxyPool = ProxyPool.getInstance(immutableConfig);
        return proxyPool.recover(n);
    }

    @UDFunction
    public static String getProxyPoolStatus() {
        ProxyPool proxyPool = ProxyPool.getInstance(immutableConfig);
        return proxyPool.toString();
    }

    @UDFunction
    public static boolean isNumeric(String str) {
        return StringUtils.isNumeric(str);
    }

    @UDFunction
    public static String getDomain(String url) {
        return URLUtil.getDomainName(url, "");
    }

    @UDFunction
    public static String re1(String text, String regex) {
        return new RegexExtractor().re1(text, regex);
    }

    @UDFunction
    public static String re1(String text, String regex, int group) {
        return new RegexExtractor().re1(text, regex, group);
    }

    @UDFunction
    public static ValueArray re2(String text, String regex) {
        Pair<String, String> result = new RegexExtractor().re2(text, regex);
        Value[] array = {ValueString.get(result.getKey()), ValueString.get(result.getValue())};
        return ValueArray.get(array);
    }

    @UDFunction
    public static ValueArray re2(String text, String regex, int keyGroup, int valueGroup) {
        Pair<String, String> result = new RegexExtractor().re2(text, regex, keyGroup, valueGroup);
        Value[] array = {ValueString.get(result.getKey()), ValueString.get(result.getValue())};
        return ValueArray.get(array);
    }

    @UDFunction
    public static ValueArray array(Value... values) {
        return ValueArray.get(values);
    }

    /**
     * Set config to the given value
     * @param h2session H2 session
     * @param name The property name to set
     * @param value The property value to set
     *              If value is null, do not set and return null
     * @return The old value or null on failure
     * */
    private static String getAndSetConf(@H2Context Session h2session, String name, @Nullable String value, int ttl) {
        Objects.requireNonNull(name);

        QuerySession session = engine.getSession(new DbSession(h2session));
        String old = session.getConfig().get(name);
        if (value != null) {
            session.getConfig().set(name, value, ttl);
            return old;
        }

        return null;
    }

    private static Duration getDuration(String duration) {
        // default unit is second
        if (StringUtils.isNumeric(duration)) {
            duration += "s";
        }

        return SParser.wrap(duration).getDuration();
    }
}

File: pulsar-ql-server/src/main/kotlin/ai/platon/pulsar/ql/h2/udfs/CommonFunctions.kt
package ai.platon.pulsar.ql.h2.udfs

import ai.platon.pulsar.common.RegexExtractor
import ai.platon.pulsar.common.SParser
import ai.platon.pulsar.common.config.CapabilityTypes.*
import ai.platon.pulsar.crawl.common.URLUtil
import ai.platon.pulsar.persist.metadata.BrowserType
import ai.platon.pulsar.persist.metadata.FetchMode
import ai.platon.pulsar.ql.QuerySession
import ai.platon.pulsar.ql.SQLContext
import ai.platon.pulsar.ql.annotation.UDFGroup
import ai.platon.pulsar.ql.annotation.UDFunction
import ai.platon.pulsar.ql.h2.H2SessionFactory
import org.apache.commons.lang3.StringUtils
import org.h2.engine.Session
import ai.platon.pulsar.ql.annotation.H2Context
import ai.platon.pulsar.ql.h2.addColumn
import org.h2.tools.SimpleResultSet
import org.h2.value.Value
import org.h2.value.ValueArray
import org.h2.value.ValueString
import org.slf4j.LoggerFactory
import java.io.File
import java.lang.Long.MAX_VALUE
import java.time.Duration
import java.util.*

@UDFGroup
object CommonFunctions {

    private val log = LoggerFactory.getLogger(CommonFunctions::class.java)

    private val sqlContext = SQLContext.getOrCreate()
    private val unmodifiedConfig = sqlContext.unmodifiedConfig

    /**
     * Set volatileConfig to the given value
     *
     * TODO: do we need FetchMode any more?
     *
     * @param h2session H2 session
     * @param mode The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     */
    @UDFunction
    @JvmStatic
    fun setFetchMode(@H2Context h2session: Session, mode_: String, ttl: Int): String? {
        var mode = mode_
        try {
            mode = FetchMode.valueOf(mode.toUpperCase()).toString()
        } catch (e: Throwable) {
            log.warn("Unknown FetchMode $mode")
            return null
        }

        log.debug("Set fetch mode to $mode")
        return getAndSetConf(h2session, FETCH_MODE, mode, ttl)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's unmodified volatileConfig")
    @JvmStatic
    fun unsetFetchMode(@H2Context h2session: Session): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(FETCH_MODE)
    }

    /**
     * Set volatileConfig to the given value
     * @param h2session H2 session
     * @param browser The browser to use
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     */
    @UDFunction
    @JvmStatic
    fun setBrowser(@H2Context h2session: Session, browser: String, ttl: Int): String? {
        val browserType = BrowserType.fromString(browser)
        if (browserType == BrowserType.NATIVE) {
            unsetFetchMode(h2session)
        } else {
            setFetchMode(h2session, FetchMode.BROWSER.name, ttl)
        }

        log.debug("Set browser to $browser")
        return getAndSetConf(h2session, BROWSER_TYPE, browserType.name, ttl)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's unmodified volatileConfig")
    @JvmStatic
    fun unsetBrowser(@H2Context h2session: Session): String? {
        val session = getSession(h2session)
        unsetFetchMode(h2session)
        return session.sessionConfig.getAndUnset(BROWSER_TYPE)
    }

    /**
     * Also retrieve faded links whenever get out links of a page
     * @param ttl The property value time to live in session
     */
    @UDFunction
    @JvmStatic
    fun enableFadedLinks(@H2Context h2session: Session, ttl: Int): String? {
        return getAndSetConf(h2session, PARSE_RETRIEVE_FADED_LINKS, true.toString(), ttl)
    }

    @UDFunction
    @JvmStatic
    fun disableFadedLinks(@H2Context h2session: Session, ttl: Int): String? {
        return getAndSetConf(h2session, PARSE_RETRIEVE_FADED_LINKS, false.toString(), ttl)
    }

    @UDFunction
    @JvmStatic
    fun setEagerFetchLimit(@H2Context h2session: Session, parallel: Boolean, ttl: Int): String? {
        return getAndSetConf(h2session, FETCH_CONCURRENCY, parallel.toString(), ttl)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's unmodified volatileConfig")
    @JvmStatic
    fun unsetEagerFetchLimit(@H2Context h2session: Session): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(FETCH_CONCURRENCY)
    }

    /**
     * Set page expires, this volatileConfig affects just in h2session scope
     * SQL examples:
     * `CALL setPageExpires('1d')`
     * `CALL setPageExpires('PT24H')`
     * `CALL setPageExpires('1s')`
     * `CALL setPageExpires('1')`
     *
     * @param h2session  The H2 session, auto injected by h2 runtime
     * @param duration The duration string.
     * Supported formats: 1. number in seconds, 2. ISO-8601 standard and 3. hadoop time duration format
     * ISO-8601 standard : PnDTnHnMn.nS
     * Hadoop time duration format : Valid units are : ns, us, ms, s, m, h, d.
     * @param ttl The property value time to live in session
     * @return The old page expires in seconds, or null if failed
     */
    @UDFunction(description = "Set the page expire time with time-to-life of the calling session")
    @JvmStatic
    @JvmOverloads
    fun setPageExpires(@H2Context h2session: Session, duration: String, ttl: Int = 1): String? {
        val d = getDuration(duration)
        val value = d?.toString()
        return getAndSetConf(h2session, STORAGE_DATUM_EXPIRES, value, ttl)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's unmodified volatileConfig")
    @JvmStatic
    fun unsetPageExpires(@H2Context h2session: Session): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(STORAGE_DATUM_EXPIRES)
    }

    /**
     * Set volatileConfig to the given value
     * @param h2session H2 session
     * @param duration The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     */
    @UDFunction(description = "Set the page load timeout with time-to-life of the calling session")
    @JvmStatic
    @JvmOverloads
    fun setPageLoadTimeout(@H2Context h2session: Session, duration: String, ttl: Int = 1): String? {
        val d = getDuration(duration)
        val value = d?.toString()
        return getAndSetConf(h2session, FETCH_PAGE_LOAD_TIMEOUT, value, ttl)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's unmodified volatileConfig")
    @JvmStatic
    fun unsetPageLoadTimeout(@H2Context h2session: Session): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(FETCH_PAGE_LOAD_TIMEOUT)
    }

    /**
     * Set volatileConfig to the given value
     * @param h2session H2 session
     * @param duration The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     */
    @UDFunction(description = "Set the script timeout with time-to-life of the calling session")
    @JvmStatic
    @JvmOverloads
    fun setScriptTimeout(@H2Context h2session: Session, duration: String, ttl: Int = 1): String? {
        val d = getDuration(duration)
        val value = d?.toString()
        return getAndSetConf(h2session, FETCH_SCRIPT_TIMEOUT, value, ttl)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's unmodified volatileConfig")
    @JvmStatic
    fun unsetScriptTimeout(@H2Context h2session: Session): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(FETCH_SCRIPT_TIMEOUT)
    }

    /**
     * Set volatileConfig to the given value
     * @param h2session H2 session
     * @param count The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     */
    @UDFunction(description = "Set the scroll down count with time-to-life of the calling session")
    @JvmStatic
    @JvmOverloads
    fun setScrollDownCount(@H2Context h2session: Session, count: Int, ttl: Int = 1): String? {
        return getAndSetConf(h2session, FETCH_SCROLL_DOWN_COUNT, count.toString(), ttl)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's initial volatileConfig")
    @JvmStatic
    fun unsetScrollDownCount(@H2Context h2session: Session): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(FETCH_SCROLL_DOWN_COUNT)
    }

    /**
     * Set volatileConfig to the given value
     * @param h2session H2 session
     * @param duration The property value to set
     * @param ttl The property value time to live in session
     * @return The old value or null on failure
     */
    @UDFunction(description = "Set the scroll interval with time-to-life of the calling session")
    @JvmStatic
    fun setScrollInterval(@H2Context h2session: Session, duration: String, ttl: Int): String? {
        val d = getDuration(duration)
        val value = d?.toString()
        return getAndSetConf(h2session, FETCH_SCROLL_DOWN_INTERVAL, value, ttl)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's initial volatileConfig")
    @JvmStatic
    fun setScrollInterval(@H2Context h2session: Session): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(FETCH_SCROLL_DOWN_INTERVAL)
    }

    /**
     * Set h2session scope configuration.
     *
     * SQL examples:
     * `CALL setConf('fetch.page.load.timeout', '1m')`
     * `CALL setConf('fetch.fetch.mode', 'NATIVE')`
     *
     * @param h2session The H2 session, auto injected by h2 runtime
     * @return The old value of the key
     */
    @UDFunction(description = "Set the volatileConfig property associated by name with time-to-life of the calling session")
    @JvmStatic
    @JvmOverloads
    fun setConfig(@H2Context h2session: Session, name: String, value: String, ttl: Int = Integer.MAX_VALUE / 2): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndSet(name, value, ttl)
    }

    @UDFunction(description = "Set the volatileConfig property associated by name with time-to-life of the calling session")
    @JvmStatic
    fun setConf(@H2Context h2session: Session, name: String, value: String, ttl: Int): String? {
        return setConfig(h2session, name, value, ttl)
    }

    @UDFunction(description = "Set the volatileConfig property associated by name of the calling session")
    @JvmStatic
    fun setConf(@H2Context h2session: Session, name: String, value: String): String? {
        return setConfig(h2session, name, value, Integer.MAX_VALUE / 2)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's initial volatileConfig")
    @JvmStatic
    fun unsetConf(@H2Context h2session: Session, name: String): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(name)
    }

    @UDFunction(description = "Unset the volatileConfig property of the calling session, " +
            "so it fallback to the process's initial volatileConfig")
    @JvmStatic
    fun unsetConfig(@H2Context h2session: Session, name: String): String? {
        val session = getSession(h2session)
        return session.sessionConfig.getAndUnset(name)
    }

    @UDFunction(description = "Get the value associated by the given key of the calling session")
    @JvmStatic
    fun getConf(@H2Context h2session: Session, name: String): String? {
        val session = getSession(h2session)
        return session.sessionConfig.get(name)
    }

    @UDFunction(description = "Get the value associated by the given key of the calling session")
    @JvmStatic
    fun getConfig(@H2Context h2session: Session, name: String): String? {
        return getConf(h2session, name)
    }

    @UDFunction(description = "Get the initial configuration properties of the process")
    @JvmStatic
    fun volatileConfig(): SimpleResultSet {
        val rs = SimpleResultSet()

        rs.addColumn("NAME")
        rs.addColumn("VALUE")

        for ((key, value) in unmodifiedConfig.unbox()) {
            rs.addRow(key, value)
        }

        return rs
    }

    @UDFunction(description = "Get the configuration properties of the calling session")
    @JvmStatic
    fun sessionConfig(@H2Context h2session: Session): SimpleResultSet {
        val rs = SimpleResultSet()
        rs.addColumn("NAME")
        rs.addColumn("VALUE")

        val session = getSession(h2session)
        for ((key, value) in session.sessionConfig.unbox()) {
            rs.addRow(key, value)
        }

        return rs
    }

    @UDFunction(description = "Get the system info")
    @JvmStatic
    fun sysInfo(@H2Context h2session: Session): SimpleResultSet {
        val rs = SimpleResultSet()
        rs.addColumn("NAME")
        rs.addColumn("VALUE")

        /* Total number of processors or cores available to the JVM */
        rs.addRow("Available processors (cores)", Runtime.getRuntime().availableProcessors())

        /* Total amount of free memory available to the JVM */
        rs.addRow("Free memory (Mbytes)", Runtime.getRuntime().freeMemory() / 1024 / 1024)

        /* This will return Long.MAX_VALUE if there is no preset limit */
        val maxMemory = Runtime.getRuntime().maxMemory()
        /* Maximum amount of memory the JVM will attempt to use */
        rs.addRow("Maximum memory (Mbytes)", if (maxMemory == MAX_VALUE) "no limit" else java.lang.Long.toString(maxMemory / 1024 / 1024))

        /* Total memory currently available to the JVM */
        rs.addRow("Total memory available to JVM (Mbytes)", Runtime.getRuntime().totalMemory() / 1024 / 1024)

        /* Get a list of all filesystem roots on this system */
        val roots = File.listRoots()
        for (root in roots) {
            /* For each filesystem root, print some info */
            rs.addRow("File system root", root.absolutePath)
            rs.addRow("Total space (Mbytes)", root.totalSpace / 1024 / 1024)
            rs.addRow("Free space (Mbytes)", root.freeSpace / 1024 / 1024)
            rs.addRow("Usable space (Mbytes)", root.usableSpace / 1024 / 1024)
        }

        return rs
    }

    @UDFunction(description = "Test if the given string is a number")
    @JvmStatic
    fun isNumeric(str: String): Boolean {
        return StringUtils.isNumeric(str)
    }

    @UDFunction(description = "Get the domain of a url")
    @JvmStatic
    fun getDomain(url: String): String {
        return URLUtil.getDomainName(url, "")
    }

    @UDFunction(description = "Extract the first group of the result of java.util.regex.matcher()")
    @JvmStatic
    fun re1(text: String, regex: String): String {
        return RegexExtractor().re1(text, regex)
    }

    @UDFunction(description = "Extract the nth group of the result of java.util.regex.matcher()")
    @JvmStatic
    fun re1(text: String, regex: String, group: Int): String {
        return RegexExtractor().re1(text, regex, group)
    }

    @UDFunction(description = "Extract two groups of the result of java.util.regex.matcher()")
    @JvmStatic
    fun re2(text: String, regex: String): ValueArray {
        val result = RegexExtractor().re2(text, regex)
        val array = arrayOf(ValueString.get(result.key), ValueString.get(result.value))
        return ValueArray.get(array)
    }

    @UDFunction(description = "Extract two groups(key and value) of the result of java.util.regex.matcher()")
    @JvmStatic
    fun re2(text: String, regex: String, keyGroup: Int, valueGroup: Int): ValueArray {
        val result = RegexExtractor().re2(text, regex, keyGroup, valueGroup)
        val array = arrayOf(ValueString.get(result.key), ValueString.get(result.value))
        return ValueArray.get(array)
    }

    @UDFunction
    @JvmStatic
    fun make_array(vararg values: Value): ValueArray {
        return ValueArray.get(values)
    }

    @UDFunction
    @JvmStatic
    fun isEmpty(array: ValueArray): Boolean {
        return array.list.isEmpty()
    }

    @UDFunction
    @JvmStatic
    fun isNotEmpty(array: ValueArray): Boolean {
        return array.list.isNotEmpty()
    }

    /**
     * Set volatileConfig to the given value
     * @param h2session H2 session
     * @param name The property name to set
     * @param value The property value to set
     * If value is null, do not set and return null
     * @return The old value or null on failure
     */
    private fun getAndSetConf(@H2Context h2session: Session, name: String, value: String?, ttl: Int): String? {
        Objects.requireNonNull(name)

        val session = getSession(h2session)
        val old = session.sessionConfig.get(name)
        if (value != null) {
            session.sessionConfig.set(name, value, ttl)
            return old
        }

        return null
    }

    private fun getSession(h2session: Session): QuerySession {
        return H2SessionFactory.getSession(h2session.serialId)
    }

    private fun getDuration(duration: String): Duration? {
        var d = duration
        // default unit is second
        if (StringUtils.isNumeric(d)) {
            d += "s"
        }

        return SParser.wrap(d).duration
    }
}
